src/python/gem5/resources/downloader.py

```diff
@@ -368,3 +380,19 @@ def safe_extract(
 
                 safe_extract(f, unpack_to)
             os.remove(download_dest)
+
+
+def _file_uri_to_path(uri: str) -> Optional[Path]:
+    """
+    Converts a file URI to a local path.
+
+    :param uri: The file URI to convert.
+
+    :returns: The path to the file.
+    """
+    if urlparse(uri).netloc == "" or urlparse(uri).netloc == "localhost":
+        local_path = urlparse(uri).path
+        if os.path.exists(local_path):
```

Comment:
So I think there was a small misunderstanding in what I requested (please let me know if there's a reason you did otherwise).

My intent was that "None" would be returned if it wasn't conforming to the File schema (e.g., it may be a https URL, for example). The exception case was only for the case that the URI was a valid file scheme URI but for a host other than localhost support.

Basically:
- None: This isn't a File scheme URI
- Exception: Is a File scheme URI but for a feature we haven't implemented.
- Path value: just the Path specified by the file URI. It doesn't need to exist. `Path.exits()` will do that for us I believe and can be done outside of the function.

Reply:
> What would be an example of a File scheme URI feature we haven't implemented?

So the File scheme URI allows for getting files on different hosts. For example `file://liberty/home/bbruce/some-file.txt` is valid. It's saying "The file '/home/bburce/some-file.txt' in the machine with hostname 'liberty'". This is something we're not going to implement. It's valid use of the standard but we're not going to implement it. That's the only thing, I think, we're not really supporting here. We're saying it must be the File scheme URL for localhost.

>After reading your input here, I am thinking the function I want to add should do the following:

> 1. sanity check if the scheme is "file".
> 2. if true, then checks the netloc
> 3. if the netloc is not localhost or blank, throws an Exception
> 4. if the netloc is localhost or blank, returns the Path object of the URI passed into the function
> 5. if the scheme being "file" is false, return None

That is what I wanted. I know we're really implementing two functions in one here: `is_file_scheme(uri: str)->bool` and `get_path_from_file_url(uri:str)->Path` by making `None` communicate the case where `is_file_scheme(uri) == False`, but this isn't a public API will almost certainly  just be used for this case, so let's keep it simple. Just make sure the docstring explains what each output actually means (you can use my suggested change).

>Meanwhile, in the get_resource function:

> 1. if the return value of the function is a Path, check if the Path exists and if yes, do the shutil.copy function as we have been doing so far.
> 2. if the return value is None, throw an exception stating that the scheme is not "file" (not sure what the behavior should be here).

(1.) makes sense to me. 

For (2.) I'll leave some comments in your code, it should be easier to explain there.